File: 01-subarray_sums_and_range_queries/SubArraySum.java
/**
 * Day 1: Subarray Sums & Range Queries
 * Java Implementation
 *
 * Author: Andres
 * Date: October 2025
 */

public class SubArraySum{

    /**
     * Main method to run tests
     */
    public static void main(String[] args) {
        System.out.println("🧪 Running DSA Solution Tests...");
        System.out.println("=".repeat(50));
        
        runTests();
    }
    
    /**
     * Simple assertion method for testing
     */
    private static void assertEquals(int actual, int expected, String testName) {
        if (actual == expected) {
            System.out.println("✅ " + testName + ": " + actual);
        } else {
            System.out.println("❌ " + testName + ": expected " + expected + ", got " + actual);
            System.exit(1);
        }
    }
    
    /**
     * Run all tests for the implemented algorithms
     */
    private static void runTests() {
        // Test cases with expected results
        TestCase[] testCases = {
            new TestCase(new int[]{1, 2, 3}, 20, "Basic example [1,2,3]"),
            new TestCase(new int[]{2, 1, 3}, 19, "Reordered [2,1,3]"),
            new TestCase(new int[]{5}, 5, "Single element [5]"),
            new TestCase(new int[]{}, 0, "Empty array"),
            new TestCase(new int[]{-1, 2, -3}, -4, "Mixed signs [-1,2,-3]"),
            new TestCase(new int[]{3, 3, 3}, 30, "Repeated elements [3,3,3]")
        };
        
        String[] algorithmNames = {
            "naive_brute_force",
            "optimized_brute_force", 
            "contribution_technique",
            "vectorized_contribution",
            "vector_half_contribution"
        };
        
        int totalTests = 0;
        int passedTests = 0;
        
        for (String algorithm : algorithmNames) {
            System.out.println("\n📋 Testing " + algorithm + ":");
            
            for (TestCase testCase : testCases) {
                try {
                    int result = runAlgorithm(algorithm, testCase.input);
                    assertEquals(result, testCase.expected, testCase.name);
                    passedTests++;
                } catch (Exception e) {
                    System.out.println("❌ " + testCase.name + ": " + e.getMessage());
                }
                totalTests++;
            }
        }
        
        System.out.println("\n📊 Results: " + passedTests + "/" + totalTests + " tests passed");
        System.out.println(passedTests == totalTests ? "🎉 All tests passed!" : "💥 Some tests failed!");
    }
    
    /**
     * Helper method to run specific algorithm by name
     */
    private static int runAlgorithm(String algorithmName, int[] arr) {
        switch (algorithmName) {
            case "naive_brute_force":
                return SubarraySum.naive_brute_force(arr);
            case "optimized_brute_force":
                return SubarraySum.optimized_brute_force(arr);
            case "contribution_technique":
                return SubarraySum.contribution_technique(arr);
            case "vectorized_contribution":
                return SubarraySum.vectorized_contribution(arr);
            case "vector_half_contribution":
                return SubarraySum.vector_half_contribution(arr);
            default:
                throw new IllegalArgumentException("Unknown algorithm: " + algorithmName);
        }
    }
    
    /**
     * Test case data structure
     */
    private static class TestCase {
        int[] input;
        int expected;
        String name;
        
        TestCase(int[] input, int expected, String name) {
            this.input = input;
            this.expected = expected;
            this.name = name;
        }
    }

    /**
     * Static inner class containing all subarray sum algorithms
     */
    public static class SubarraySum {

        /**
         * Naive brute force approach - O(n³) time complexity
         * Generates all subarrays and calculates their sums
         */
        public static int naive_brute_force(int[] arr) {
            int totalSum = 0;
            int n = arr.length;

            // Triple nested loop to generate all subarrays
            for (int i = 0; i < n; i++) {              // Start index
                for (int j = i; j < n; j++) {          // End index
                    for (int k = i; k <= j; k++) {     // Sum elements in subarray [i,j]
                        totalSum += arr[k];
                    }
                }
            }
            return totalSum;
        }

        /**
         * Optimized brute force approach - O(n²) time complexity
         * Uses running sum to avoid recalculating subarray sums
         */
        public static int optimized_brute_force(int[] arr) {
            int n = arr.length;
            int totalSum = 0;

            for (int i = 0; i < n; i++) {
                int currentSum = 0;
                for (int j = i; j < n; j++) {
                    currentSum += arr[j];     // Add current element to running sum
                    totalSum += currentSum;   // Add this subarray sum to total
                }
            }
            return totalSum;
        }

        /**
         * Contribution technique - O(n) time complexity
         * Calculates how many times each element appears in all subarrays
         */
        public static int contribution_technique(int[] arr) {
            int n = arr.length;
            int totalSum = 0;
            
            for (int k = 0; k < n; k++) {
                // Element at index k appears in (k+1) * (n-k) subarrays
                int contribution = (k + 1) * (n - k);
                totalSum += arr[k] * contribution;
            }
            return totalSum;
        }

        /**
         * Vectorized contribution approach - O(n) time complexity
         * Precomputes contribution values then calculates final sum
         */
        public static int vectorized_contribution(int[] arr) {
            int n = arr.length;
            
            if (n == 0) return 0;
            
            // Precompute contribution values for each position
            int[] contributions = new int[n];
            for (int i = 0; i < n; i++) {
                contributions[i] = (i + 1) * (n - i);
            }
            
            // Calculate total sum using contributions
            int totalSum = 0;
            for (int i = 0; i < n; i++) {
                totalSum += arr[i] * contributions[i];
            }
            
            return totalSum;
        }
        
        /**
         * Vector half contribution approach - O(n) time complexity
         * Uses symmetric property of contributions to compute only half
         */
        public static int vector_half_contribution(int[] arr) {
            int n = arr.length;
            
            if (n == 0) return 0;
            
            int m = (n + 1) / 2; // ceiling division
            
            // Compute first half of contributions
            int[] firstHalf = new int[m];
            for (int k = 0; k < m; k++) {
                firstHalf[k] = (k + 1) * (n - k);
            }
            
            // Create mirror based on array length parity
            int[] contributions = new int[n];
            System.arraycopy(firstHalf, 0, contributions, 0, m);
            
            if (n % 2 == 0) {
                // Even length: mirror entire first half
                for (int i = 0; i < m; i++) {
                    contributions[m + i] = firstHalf[m - 1 - i];
                }
            } else {
                // Odd length: mirror all except center element
                for (int i = 0; i < m - 1; i++) {
                    contributions[m + i] = firstHalf[m - 2 - i];
                }
            }
            
            // Calculate total sum using contributions
            int totalSum = 0;
            for (int i = 0; i < n; i++) {
                totalSum += arr[i] * contributions[i];
            }
            
            return totalSum;
        }
    }
}

File: 01-subarray_sums_and_range_queries/subArraySum.js
#!/usr/bin/env node
/**
 * Day 1: Subarray Sums & Range Queries
 * JavaScript/Node.js Implementation
 *
 * Author: Andres
 * Date: October 2025
 */

// TODO: Implement solutions for both problems
// naive_brute_force

function naive_brute_force(arr) {
  const n = arr.length;
  let total_sum = 0;

  for (let i = 0; i < n; i++) {
    for (let j = i; j < n; j++) {
      let subarray_sum = 0;
      for (let k = i; k <= j; k++) {
        subarray_sum += arr[k];
      }
      total_sum += subarray_sum;
    }
  }
  return total_sum;
}

function optimized_brute_force(arr) {
  const n = arr.length;
  let total_sum = 0;

  for (let i = 0; i < n; i++) {
    let current_subarray_sum = 0;

    for (let j = i; j < n; j++) {
      current_subarray_sum += arr[j];
      total_sum += current_subarray_sum;
    }
  }
  return total_sum;
}

function contribution_technique(arr) {
  const n = arr.length;
  let contribution = (k, n) => (k + 1) * (n - k);
  let total_sum = 0;
  for (let k = 0; k < n; k++) {
    total_sum += arr[k] * contribution(k, n);
  }
  return total_sum;
}

// vectorized_contribution
function vectorized_contribution(arr) {
  const n = arr.length;
  const contribution = (k, n) => (k + 1) * (n - k);
  let contributions = [];
  for (let k = 0; k < n; k++) {
    contributions.push(contribution(k, n));
  }
  let total_sum = 0;
  for (let i = 0; i < n; i++) {
    total_sum += arr[i] * contributions[i];
  }
  return total_sum;
}

// vector_half_contribution
function vector_half_contribution(arr) {
  const n = arr.length;
  const contribution = (k, n) => (k + 1) * (n - k);
  const m = Math.ceil(n / 2);
  // compute half of the contribution array
  let first_half = [];
  for (let k = 0; k < m; k++) {
    first_half.push(contribution(k, n));
  }
  let mirror =
    n % 2 == 0
      ? first_half.slice().reverse()
      : first_half.slice(0, m - 1).reverse();

  let contributions = first_half.concat(mirror);
  let total_sum = 0;
  for (let i = 0; i < n; i++) {
    total_sum += arr[i] * contributions[i];
  }
  return total_sum;
}

// m = (n + 1) // 2 ; // ceiling division for midpoint

// Export functions for testing
module.exports = {
  naive_brute_force,
  optimized_brute_force,
  contribution_technique,
  vectorized_contribution,
  vector_half_contribution
};

// Run tests if this file is executed directly
if (require.main === module) {
  // Simple assertion-based tests
  const assert = (actual, expected, description) => {
    if (actual === expected) {
      console.log(`✅ ${description}: ${actual}`);
    } else {
      console.log(`❌ ${description}: expected ${expected}, got ${actual}`);
      process.exit(1);
    }
  };

  const algorithms = [
    { name: 'naive_brute_force', func: naive_brute_force },
    { name: 'optimized_brute_force', func: optimized_brute_force },
    { name: 'contribution_technique', func: contribution_technique },
    { name: 'vectorized_contribution', func: vectorized_contribution },
    { name: 'vector_half_contribution', func: vector_half_contribution }
  ];

  const testCases = [
    { input: [1, 2, 3], expected: 20, name: "Basic example [1,2,3]" },
    { input: [2, 1, 3], expected: 19, name: "Reordered [2,1,3]" },
    { input: [5], expected: 5, name: "Single element [5]" },
    { input: [], expected: 0, name: "Empty array" },
    { input: [-1, 2, -3], expected: -4, name: "Mixed signs [-1,2,-3]" },
    { input: [3, 3, 3], expected: 30, name: "Repeated elements [3,3,3]" }
  ];

  console.log('🧪 Running DSA Solution Tests...');
  console.log('='.repeat(50));
  
  let totalTests = 0;
  let passedTests = 0;
  
  algorithms.forEach(algorithm => {
    console.log(`\n📋 Testing ${algorithm.name}:`);
    testCases.forEach(testCase => {
      try {
        const result = algorithm.func(testCase.input);
        assert(result, testCase.expected, `${testCase.name}`);
        passedTests++;
      } catch (error) {
        console.log(`❌ ${testCase.name}: ${error.message}`);
      }
      totalTests++;
    });
  });
  
  console.log(`\n📊 Results: ${passedTests}/${totalTests} tests passed`);
  console.log(passedTests === totalTests ? '🎉 All tests passed!' : '💥 Some tests failed!');
}
